kwok/cloudprovider/helpers.go

```diff
@@ -30,78 +32,102 @@ import (
-	var instanceTypes []*cloudprovider.InstanceType
-	for _, cpu := range []int{1, 2, 4, 8, 16, 32, 48, 64, 96, 128, 192, 256} {
-		for _, memFactor := range []int{2, 4, 8} {
-			for _, os := range []sets.Set[string]{sets.New(string(v1.Linux)), sets.New(string(v1.Windows))} {
-				for _, arch := range []string{v1beta1.ArchitectureAmd64, v1beta1.ArchitectureArm64} {
-					// Construct instance type details, then construct offerings.
-					mem := cpu * memFactor
-					pods := lo.Clamp(cpu*16, 0, 1024)
-					labels := MakeInstanceTypeLabels(cpu, memFactor)
-					opts := InstanceTypeOptions{
-						Name:             fmt.Sprintf("%s-%s-%s-%s", labels[InstanceFamilyLabelKey], labels[InstanceSizeLabelKey], arch, strings.Join(sets.List(os), ",")),
-						Architecture:     arch,
-						OperatingSystems: os,
-						Resources: v1.ResourceList{
-							v1.ResourceCPU:              resource.MustParse(fmt.Sprintf("%d", cpu)),
-							v1.ResourceMemory:           resource.MustParse(fmt.Sprintf("%dGi", mem)),
-							v1.ResourcePods:             resource.MustParse(fmt.Sprintf("%d", pods)),
-							v1.ResourceEphemeralStorage: resource.MustParse("20G"),
-						},
-						InstanceTypeLabels: labels,
-					}
-					price := PriceFromResources(opts.Resources)
-
-					opts.Offerings = cloudprovider.Offerings{}
-					for _, zone := range KwokZones {
-						for _, ct := range []string{v1beta1.CapacityTypeSpot, v1beta1.CapacityTypeOnDemand} {
-							opts.Offerings = append(opts.Offerings, cloudprovider.Offering{
-								CapacityType: ct,
-								Zone:         zone,
-								Price:        lo.Ternary(ct == v1beta1.CapacityTypeSpot, price*.7, price),
-								Available:    true,
-							})
-						}
-					}
-					instanceTypes = append(instanceTypes, newInstanceType(opts))
-				}
-			}
+// parseSizeFromType will attempt to discover the instance size if it matches a special AWS format.
+// If not, fall back to the cpu value.  This works for both Azure and GCP (and the "generic" instances
+// generated by tools/gen_instances.go)
+func parseSizeFromType(ty, cpu string) string {
+	if matches := awsRegexp.FindStringSubmatch(ty); matches != nil {
+		return matches[1]
+	}
+
+	return cpu
+}
+
+// parseFamilyFromType will attempt to discover the instance family from the node type.  The three major cloud providers
+// have the following formats:
+//
+// AWS   - [Family].[Size]
+// GCP   - [Family]-[Configuration]-[# of vCPUs])
+// Azure - [Family] + [Sub-family]* + [# of vCPUs] + ... + [Version]
+//
+// So here, we split on [.-], and if that fails, fall back to the first character of the instance type name.
+func parseFamilyFromType(ty string) string {
+	familySplit := familyDelim.Split(ty, 2)
+	if len(familySplit) < 2 {
+		return ty[0:1]
+	}
+	return familySplit[0]
+}
+
+func setDefaultOptions(opts InstanceTypeOptions) InstanceTypeOptions {
+	var cpu, memory string
+	for res, q := range opts.Resources {
+		switch res {
+		case "cpu":
```

Comment:
```suggestion
		case v1.ResourceCPU:
```

consider using the upstream names for these resource labels